xbmc/cores/AudioEngine/Sinks/AESinkPULSE.cpp

```diff
@@ -1089,22 +1097,37 @@ unsigned int CAESinkPULSE::AddPackets(uint8_t **data, unsigned int frames, unsig
   pa_threaded_mainloop_lock(m_MainLoop);
 
   unsigned int available = frames * m_format.m_frameSize;
-  unsigned int length = 0;
+  unsigned int length = m_periodSize;
   void *buffer = data[0]+offset*m_format.m_frameSize;
-  // care a bit for fragmentation
-  while ((length = pa_stream_writable_size(m_Stream)) < m_periodSize)
+  unsigned int wait_time = static_cast<unsigned int>(1000.0 * m_BufferSize / m_BytesPerSecond);
+  m_extTimer.Set(wait_time);
+  // we don't want to block forever - if timer expires pa_stream_write will
+  // fail - therefore we don't care and just return 0;
+  while (!m_extTimer.IsTimePast())
+  {
+    if (m_requestedBytes > 0)
+      break;
     pa_threaded_mainloop_wait(m_MainLoop);
+  }
 
-  length =  std::min(length, available);
+  if (m_extTimer.IsTimePast())
+  {
+    CLog::Log(LOGERROR, "Sink Timer expired for more than buffer time: %ul", wait_time);
```

Comment:
you should return 0 here

Reply:
Okay - plus unlocking the mainloop then.